### ISSUE TITLE
Human readable chaincode address aliases

### ISSUE THREAD

**Issue by markparz** [hyperledger-archives/fabric#1077]
The current Chaincode address is now a hash string, which is impossible to read. It would be much better if the hash string can be mapped to a human readable alias.

What we have today is a low level unique identifier generated and returned by the fabric automatically. This is good as a low level implementation detail but we need higher level mechanisms for allowing users to name chaincode. The naming service would provide such a mechanism. This can be tied to a naming scheme that includes the validator name or DNS name etc.

This would allow a query to find the asset my business wants to trade (otherwise such mapping has to be defined locally)

**Comment by smithbk** [comment]
The way in which cloud foundry handles this is that each object in the system has 3 things:
1) name, 2) type, and 3) guid.  The name and type are of course provided by the object creator and the guid is generated like a hash.
At any point in time, the <name,type> combination must be unique, while guids are globally unique across time.  So if an object with name=foo and type=bar is created, deleted, and recreated, it's 2nd instance will have a different guid even if the contents are identical.
Objects can be retrieved by name & type ... or by guid.
This design has worked well for cloud foundry / bluemix as it enables someone to reference a specific object in time by using it's guid as well as referencing it by name in case it is deleted and recreated, or simply updated.

**Comment by ibmmark** [comment]
@muralisrini can you put some more detailed thought here?

**Comment by kletkeman** [comment]
In #1127, redirect of messages (events) for contract A to contract A.prime (upgraded) is handled by "naming the A.prime with the same name as A" ... is this task therefore planning on changing the prefix used for state to something that relates to the alias and not the UUID? The reason I ask is that properties are stored in buckets with the key being UUID<key from contract> and the contract A.prime only gets the same alias and not the same UUID. 

Basically, I'd like to see the database subsumed by the new contract without a copy step if possible. At least as an option.

**Comment by ghaskins** [comment]
Related to #748

**Comment by kletkeman** [comment]
As I mention in #748, I'd like to see some notion of database sharing, which really means _key prefix sharing_ for upgrade cases (take over the data with no time lag) and possible scenarios where two chaincodes sharing the same data partition (as in prefix) makes sense ...

**Comment by kletkeman** [comment]
[#1217](https://github.com/hyperledger/fabric/issues/1217) is needed for fast upgrades

**Comment by markparz** [comment]
Migrated to Jira [FAB-18](https://jira.hyperledger.org/browse/FAB-18).
Closing this issue

#1217 and #748 are being opened as well by Clayton and hung to the above epic.
